File: src/gameControl/GameEngine.js
const adb = require('adbkit');
const fs = require('fs').promises;
const path = require('path');
const TouchController = require('./controllers/TouchController');
const AutoBattleStrategy = require('./controllers/AutoBattleStrategy');
const OCRUtils = require('./ocrPrediction/OCRUtils');
const ImageRecognition = require('./ImageRecognition');

class GameEngine {
  constructor(strategyFile, adbDevice = null) {
    this.adbDevice = adbDevice;
    this.touchController = null;
    this.battleStrategy = new AutoBattleStrategy(strategyFile);
    this.ocrUtils = new OCRUtils();
    this.imageRecognition = new ImageRecognition();
    this.stopAutomation = false;
    this.screenWidth = 0;
    this.screenHeight = 0;
    this.emulatorSettings = null;
    this.client = adb.createClient();
  }

  async initialize() {
    await this.loadEmulatorSettings();
    await this.detectEmulators();
    await this.selectEmulator(this.adbDevice);
    await this.updateScreenSize();
  }

  async loadEmulatorSettings() {
    const data = await fs.readFile(path.join(__dirname, '../../config/emulator_settings.json'), 'utf8');
    this.emulatorSettings = JSON.parse(data);
  }

  async detectEmulators() {
    const devices = await this.client.listDevices();
    this.emulatorSettings.emulators = devices.map(device => ({
      name: `Emulator_${device.id}`,
      serial: device.id,
      status: device.type
    }));
  }

  async selectEmulator(adbDevice) {
    if (adbDevice) {
      const emulator = this.emulatorSettings.emulators.find(e => e.serial === adbDevice);
      if (emulator) {
        this.adbDevice = adbDevice;
      } else {
        throw new Error(`Specified ADB device ${adbDevice} not found`);
      }
    } else {
      if (this.emulatorSettings.emulators.length > 0) {
        this.adbDevice = this.emulatorSettings.emulators[0].serial;
      } else {
        throw new Error("No ADB devices detected");
      }
    }
    
    this.touchController = new TouchController(this.adbDevice);
  }

  async updateScreenSize() {
    const output = await this.client.shell(this.adbDevice, 'wm size');
    const size = output.toString().trim().split(':')[1].trim().split('x');
    [this.screenWidth, this.screenHeight] = size.map(Number);
  }

  async captureScreen() {
    const screenshotBuffer = await this.client.screencap(this.adbDevice);
    return screenshotBuffer;
  }

  async findGameObject(templatePath) {
    const screen = await this.captureScreen();
    return this.imageRecognition.templateMatching(screen, templatePath);
  }

  async waitForGameObject(templatePath, timeout = 30000) {
    const startTime = Date.now();
    while (Date.now() - startTime < timeout) {
      const result = await this.findGameObject(templatePath);
      if (result.length > 0) {
        return true;
      }
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    return false;
  }

  async performActionSequence(actions) {
    for (const action of actions) {
      if (action.type === 'tap') {
        await this.touchController.tap(action.x, action.y);
      } else if (action.type === 'wait') {
        await new Promise(resolve => setTimeout(resolve, action.duration));
      } else if (action.type === 'find_and_tap') {
        const objects = await this.findGameObject(action.template);
        if (objects.length > 0) {
          const [x, y, w, h] = objects[0];
          await this.touchController.tap(x + w/2, y + h/2);
        }
      }
    }
  }

  async detectGameState() {
    // Implement game state detection logic
    // This is a placeholder and should be implemented based on your game's specific requirements
    return 'unknown';
  }

  async getResourceValue(resource) {
    // Implement resource value retrieval logic
    // This is a placeholder and should be implemented based on your game's specific requirements
    return 0;
  }

  getEmulatorInfo() {
    return this.emulatorSettings.emulators.find(e => e.serial === this.adbDevice);
  }

  async startEmulator() {
    console.log("Starting emulator...");
    // Implement the logic to start the emulator
    // This might involve running a shell command or using an emulator-specific API
  }

  async runGameLoop() {
    while (!this.stopAutomation) {
      try {
        const screen = await this.captureScreen();
        const text = await this.ocrUtils.recognizeText(screen);
        
        if (text.includes("战斗")) {
          await this.battleStrategy.executeStrategy("battle", this);
        } else if (text.includes("Boss")) {
          await this.battleStrategy.executeStrategy("boss_battle", this);
        } else {
          await this.battleStrategy.executeStrategy("default", this);
        }
        
        await new Promise(resolve => setTimeout(resolve, 1000));
      } catch (error) {
        console.error("Error in game loop:", error);
      }
    }
  }
}

module.exports = GameEngine;

File: game_control/game_engine.py
import cv2
import numpy as np
import subprocess
import json
import os
import time
import re
import asyncio
import traceback
from .controllers.touch_controller import TouchController
from .controllers.auto_battle_strategy import AutoBattleStrategy
from .ocr_prediction.ocr_utils import OCRUtils
from .image_recognition import ImageRecognition
from utils.logger import setup_logger, log_exception
from utils.data_handler import DataHandler

class GameEngine:
    def __init__(self, strategy_file, adb_device=None):
        self.logger = setup_logger()
        try:
            self.adb_device = adb_device
            self.touch_controller = None
            self.battle_strategy = AutoBattleStrategy(strategy_file)
            self.ocr_utils = OCRUtils()
            self.image_recognition = ImageRecognition()
            self.data_handler = DataHandler()
            self.stop_automation = False
            self.screen_width = 0
            self.screen_height = 0
            self.load_emulator_settings()
            self.detect_emulators()
            self.select_emulator(adb_device)
            self.update_screen_size()
        except Exception as e:
            log_exception(self.logger, e)
            raise

    def load_emulator_settings(self):
        try:
            with open('config/emulator_settings.json', 'r') as f:
                self.emulator_settings = json.load(f)
        except Exception as e:
            self.logger.error(f"Error loading emulator settings: {str(e)}")
            raise

    def detect_emulators(self):
        try:
            result = subprocess.run(['adb', 'devices'], capture_output=True, text=True)
            lines = result.stdout.strip().split('\n')[1:]
            self.emulator_settings['emulators'] = []
            for line in lines:
                if '\t' in line:
                    serial, status = line.split('\t')
                    self.emulator_settings['emulators'].append({
                        "name": f"Emulator_{serial}",
                        "serial": serial,
                        "status": status
                    })
        except Exception as e:
            self.logger.error(f"Error detecting emulators: {str(e)}")
            raise

    def select_emulator(self, adb_device):
        try:
            if adb_device:
                emulator = next((e for e in self.emulator_settings['emulators'] if e['serial'] == adb_device), None)
                if emulator:
                    self.adb_device = adb_device
                else:
                    raise ValueError(f"指定的 ADB 设备 {adb_device} 未找到")
            else:
                if self.emulator_settings['emulators']:
                    self.adb_device = self.emulator_settings['emulators'][0]['serial']
                else:
                    raise ValueError("未检测到任何 ADB 设备")
            
            self.touch_controller = TouchController(self.adb_device)
        except Exception as e:
            self.logger.error(f"Error selecting emulator: {str(e)}")
            raise

    def update_screen_size(self):
        try:
            output = subprocess.check_output(['adb', '-s', self.adb_device, 'shell', 'wm', 'size'])
            size = output.decode().strip().split(':')[-1].strip().split('x')
            self.screen_width, self.screen_height = map(int, size)
        except Exception as e:
            self.logger.error(f"Error updating screen size: {str(e)}")
            raise

    async def run_game_loop(self):
        session_start = time.time()
        session_id = self.data_handler.save_game_session(session_start, None, None)
        
        try:
            while not self.stop_automation:
                try:
                    screen = self.capture_screen()
                    text = self.ocr_utils.recognize_text(screen)
                    
                    if "战斗" in text:
                        await self.battle_strategy.execute_strategy("battle", self)
                        self.data_handler.save_game_event(session_id, "battle", {"type": "normal"})
                    elif "Boss" in text:
                        await self.battle_strategy.execute_strategy("boss_battle", self)
                        self.data_handler.save_game_event(session_id, "battle", {"type": "boss"})
                    else:
                        await self.battle_strategy.execute_strategy("default", self)
                    
                    await asyncio.sleep(1)
                except Exception as e:
                    log_exception(self.logger, e)
                    self.logger.warning("Continuing game loop despite error")
        finally:
            session_end = time.time()
            session_duration = int(session_end - session_start)
            self.data_handler.save_game_session(session_start, session_end, session_duration)

    def capture_screen(self):
        try:
            subprocess.run(['adb', '-s', self.adb_device, 'shell', 'screencap', '-p', '/sdcard/screen.png'])
            subprocess.run(['adb', '-s', self.adb_device, 'pull', '/sdcard/screen.png', 'screen.png'])
            return cv2.imread('screen.png')
        except Exception as e:
            self.logger.error(f"Error capturing screen: {str(e)}")
            raise

    async def find_game_object(self, template_path):
        try:
            screen = self.capture_screen()
            return self.image_recognition.template_matching(screen, template_path)
        except Exception as e:
            self.logger.error(f"Error finding game object: {str(e)}")
            raise

    async def wait_for_game_object(self, template_path, timeout=30):
        start_time = time.time()
        while time.time() - start_time < timeout:
            if await self.find_game_object(template_path):
                return True
            await asyncio.sleep(1)
        return False

    async def perform_action_sequence(self, actions):
        for action in actions:
            if action['type'] == 'tap':
                self.touch_controller.tap(action['x'], action['y'])
            elif action['type'] == 'wait':
                await asyncio.sleep(action['duration'])
            elif action['type'] == 'find_and_tap':
                objects = await self.find_game_object(action['template'])
                if objects:
                    x, y, w, h = objects[0]
                    self.touch_controller.tap(x + w//2, y + h//2)

    async def detect_game_state(self):
        # 实现游戏状态检测逻辑
        pass

    async def get_resource_value(self, resource):
        # 实现资源值获取逻辑
        pass

    def get_emulator_info(self):
        return next((e for e in self.emulator_settings['emulators'] if e['serial'] == self.adb_device), None)

    def start_emulator(self):
        # 实现启动模拟器的逻辑
        self.logger.info("Starting emulator...")
        # 这里需要添加实际启动模拟器的代码

if __name__ == "__main__":
    try:
        engine = GameEngine('config/strategies.json')
        print(f"检测到的模拟器: {engine.emulator_settings['emulators']}")
        print(f"选择的模拟器: {engine.adb_device}")
        print(f"屏幕尺寸: {engine.screen_width}x{engine.screen_height}")
    except Exception as e:
        print(f"Error in main: {str(e)}")
        print(traceback.format_exc())